Name of receiver disappears #3204

bamboo3250 opened the pull request:
Fixes #3204 

![screenshot_13](https://cloud.githubusercontent.com/assets/8078323/6884534/dbd6b48c-d631-11e4-937c-8e1f8b02df6d.png)

In feedbackSubmissionsEdit.js, the value of __curSelectedOption__ and __selectedOption__ may contain character ', which will make javascript confused.

for example: curSelectedOption = __team 1 '__
then the part __"option[value='"+curSelectedOption+"']"__ becomes __"option[value='team 1 '']"__

damithc commented on src/main/webapp/js/feedbackSubmissionsEdit.js:
```diff
@@ -416,6 +416,26 @@ function validateConstSumQuestions(){
+function replaceAll(regex, replace, str) {
+	return str.replace(new RegExp(escapeRegExp(regex), 'g'), replace);
+}
+
```
I would like to see unit tests for the above two functions. (We use QUnit to javascript unit tests)

LowWeiLin commented:
The sanitization should not be done client-side. Is it possible to do it in the jsp/java files instead? Also, if the option is not a team, is the sanitization done as well(more than once)?

bamboo3250 commented:
@damithc I removed those 2 functions and made use of sanitizeForJs function.
@LowWeiLin __" Also, if the option is not a team, is the sanitization done as well(more than once)?"__
in that case, it should be either student or instructor and we use email address as the key for them. Therefore, there will be no difference between before and after applying __sanitizeForJs__ because characters such as ' and \ cannot be in the email address.

bamboo3250 commented:
Just to extend my answer, String variables are usually stored as sanitized string (for HTML). Here I use sanitization for Javascript and it still works for sanitized strings. Strings such as _&#39;_ or _'_ make no difference.

LowWeiLin commented on src/main/webapp/js/feedbackSubmissionsEdit.js:
```diff
@@ -447,7 +447,7 @@ function formatRecipientLists(){
         }
         if (curSelectedOption != "") {
-            $("select[name|="+FEEDBACK_RESPONSE_RECIPIENT+"-"+questionNumber+"]").not(this).
+        	$("select[name|="+FEEDBACK_RESPONSE_RECIPIENT+"-"+questionNumber+"]").not(this).
```
Is this change related? If not, please undo the changes.

bamboo3250 replied:
I accidently replaced 4 spaces by a tab. I will change it back when adding test case for this issue.

bamboo3250 commented:
@LowWeiLin I think we cannot fix this issue just by sanitizing from outside of Javascript. Here is the reason why:

![screenshot_1](https://cloud.githubusercontent.com/assets/8078323/6992113/d6fd5452-daf5-11e4-8d81-d5f2f0b63480.png)
In __feedbackSubmissionEdit.js__, if __curSelectedOption__'s value is __Team>' 1__, Javascript will fail at the line 452 because the find() function had __option[value='Team>' 1']__ as parameter. (there are 3 quotes).
So I sanitize it (in .java file) then __curSelectedOption__ now is __Team>\' 1__
![screenshot_2](https://cloud.githubusercontent.com/assets/8078323/6992127/9787e30e-daf6-11e4-9216-42fee6e9f865.png)
But then the find() function I mentioned just now will only see it as __Team>' 1__ (without the slash \) due to the escape character. As a result, it cannot find the option with the value __Team>\' 1__ in the HTML file.

So my point is
If __curSelectedOption__ is __Team>' 1__, Javascript will fail
On the other hand, if __curSelectedOption__ is __Team>\' 1__, javascript will not fail but cannot find what we want. 
Even if __curSelectedOption__ is __Team>\\\' 1__, there is no change.

I also tried to use sanitization for HTML but it doesnt work. So I think we must have a sanitize function inside javascript like I did before.

LowWeiLin commented:
Okay sure. Can do it with javascript and test with the cases you mentioned.

bamboo3250 commented:
Fixed main bug:
1. Added function __sanitizeForJs(string)__ in __common.js__
2. Fixed #3204 by using the function __sanitizeForJs(string)__ in __feedbackSubmissionsEdit.js__
3. Added test cases for the function __sanitizeForJs(string)__ in __CommonJsTest.js__

Added test case for the bug:
1. Changed team name in test data from __Team> 1__ to __Team >'"< 1__ in __StudentFeedbackSubmitPageUiTest.json__
2. Changed output files of the test case including:
        - __studentFeedbackSubmitPageFullyFilled.html__
        - __studentFeedbackSubmitPageModified.html__
        - __studentFeedbackSubmitPageOpen.html__
        - __studentFeedbackSubmitPagePartiallyFilled.html__
        - __unregisteredStudentFeedbackSubmitPageOpen.html__
        - __unregisteredStudentFeedbackSubmitPagePartiallyFilled.html__
3. Changed team name in the test driver __StudentFeedbackSubmitPageUiTest.java__
4. Here I found out that Selenium using XPath also got problem when handling single quote and double quote character so I added clearStringForXPath(String text) in __Sanitizer.java__
5. Used the function clearStringForXPath() in __FeedbackSubmitPage.java__ to pass the test
6. Added test case for clearStringForXPath(String text) in __SanitizerTest.java__
7. Used sanitizeForHtml in __HtmlHelper.java__ because I found out the data is not sanitized, possibly for injection and it cannot pass the test without doing this.
8. Added sanitizedForHtml in __feedbackSubmissionEdit.jsp__ as well because the data there is also not sanitized and it also cannot pass the test.
9. Dev green

LowWeiLin commented on src/main/java/teammates/common/util/Sanitizer.java:
```diff
@@ -175,4 +175,25 @@ public static String sanitizeForCsv(String str){
         return ((string == null) ? null : string.trim());
     }
+    
+    public static String clearStringForXPath(String text){
```
Could you add comments on what this function does?

LowWeiLin commented on src/main/java/teammates/common/util/Sanitizer.java:
```diff
@@ -175,4 +175,25 @@ public static String sanitizeForCsv(String str){
+        String result = "";
+        int startPos = 0;
+        for(int i=0;i<text.length();i++){
+            while((i<text.length()) && (text.charAt(i)!='\'')) i++;
```
Use curly braces for loops, according to the coding style.

LowWeiLin commented:
If thee is a problem using the xpath, why can't we use ids to reference the web element instead?

bamboo3250 commented:
![screenshot_1](https://cloud.githubusercontent.com/assets/8078323/7043618/01391962-de26-11e4-96db-64e8bbeb5b96.png)

These are radio buttons in HTML. We can see that there is no Id for these options. Therefore, the test unit has to use their values as ids.

![screenshot_2](https://cloud.githubusercontent.com/assets/8078323/7043691/7786e9f0-de26-11e4-8901-beef85bae6eb.png)

That's why the problem will raise when the team name contains those special character. Adding ids to these options is feasible but requires a lot of changes. So I decided to fix the XPath's problem instead.
